rts/motoko-rts/src/gc/incremental.rs

```diff
@@ -1,18 +1,424 @@
+const INCREMENT_ALLOCATION_FACTOR: usize = 10; // Additional time factor per concurrent allocation.
+
+// Performance note: Storing the phase-specific state in the enum would be nicer but it is much slower.
+#[derive(PartialEq)]
+enum Phase {
+    Pause,    // Inactive, waiting for the next GC run.
+    Mark,     // Incremental marking.
+    Evacuate, // Incremental evacuation compaction.
+    Update,   // Incremental pointer updates.
+    Stop,     // GC stopped on canister upgrade.
+}
+
+pub struct State {
+    phase: Phase,
+    partitioned_heap: PartitionedHeap,
+    allocation_count: usize, // Number of allocations during an active GC run.
+    mark_state: Option<MarkState>,
+    iterator_state: Option<PartitionedHeapIterator>,
+}
+
+/// GC state retained over multiple GC increments.
+static mut STATE: RefCell<State> = RefCell::new(State {
+    phase: Phase::Pause,
+    partitioned_heap: UNINITIALIZED_HEAP,
+    allocation_count: 0,
+    mark_state: None,
+    iterator_state: None,
+});
+
+/// Incremental GC.
+/// Each GC call has its new GC instance that shares the common GC state `STATE`.
+pub struct IncrementalGC<'a, M: Memory> {
+    mem: &'a mut M,
+    state: &'a mut State,
+    time: BoundedTime,
+}
+
+impl<'a, M: Memory + 'a> IncrementalGC<'a, M> {
+    /// (Re-)Initialize the entire incremental garbage collector.
+    /// Called on a runtime system start with incremental GC and also during RTS testing.
+    pub unsafe fn initialize(mem: &'a mut M, heap_base: usize) {
+        let state = STATE.get_mut();
+        state.phase = Phase::Pause;
+        state.partitioned_heap = PartitionedHeap::new(mem, heap_base);
+        state.allocation_count = 0;
+        state.mark_state = None;
+        state.iterator_state = None;
+    }
+
+    /// Each GC schedule point can get a new GC instance that shares the common GC state.
+    /// This is because the memory implementation is not stored as global variable.
+    pub unsafe fn instance(mem: &'a mut M, state: &'a mut State) -> IncrementalGC<'a, M> {
+        debug_assert!(state.partitioned_heap.is_initialized());
+        let limit = usize::saturating_add(
+            INCREMENT_BASE_LIMIT,
+            usize::saturating_mul(state.allocation_count, INCREMENT_ALLOCATION_FACTOR),
+        );
+        state.allocation_count = 0;
+        let time = BoundedTime::new(limit);
+        IncrementalGC { mem, state, time }
+    }
+
+    /// Regular GC increment invoked when the call stack is guaranteed to be empty.
+    /// As the GC cannot scan or use write barriers on the call stack, we need to ensure:
+    /// * The mark phase can only be started on an empty call stack.
+    /// * The update phase can only be completed on an empty call stack.
+    pub unsafe fn empty_call_stack_increment(&mut self, roots: Roots) {
+        assert!(self.state.phase != Phase::Stop);
+        if self.pausing() {
+            self.start_marking(roots);
+        }
+        self.increment();
```

Comment:
Any reason this call to self.increment() is outside the conditional, but the others are not?

Reply:
The motivation is to unconditionally perform at least one increment regardless of the phase (after the GC has been started). However, if there are pending GC phase switches, subsequent increments are triggered to possibly continue the next phase. 

I noticed a slight performance gain against the better readable and more symmetric version:
```
if self.pausing() {
  self.start_marking(roots);
}
if self.state.phase == Phase::Mark {
  self.mark_increment();
}
if self.mark_completed() {
  self.start_evacuating();
}
if self.state.phase == Phase::Mark {
  self.evacuation_increment();
}
if self.evacuation_completed() {
  self.start_updating();
}
if self.state.phase == Phase::Update {
  self.update_increment();
}
if self.updating_completed() {
  self.complete_run(roots);
}
```
I will remeasure the performance difference.

Reply:
**Update**: I remeasured and the performance seems the same. So, I refactored to code to have the better readable and symmetric version (mentioned above). Thank you very much for this question.